### PR TITLE
Improvement of interface post processor

### LINKED ISSUE #739: HistorySetSampling does not handle the HistorySet correctly
--------
Issue Description
--------
##### What did you expect to see happen?
For given HistorySet, the code should run successfully. 

##### What did you see instead?
```
The code crashed:
Traceback (most recent call last):
  File "/Users/wangc/projects/wangc/raven/framework/Driver.py", line 281, in <module>
    raven()
  File "/Users/wangc/projects/wangc/raven/framework/Driver.py", line 234, in raven
    simulation.run()
  File "/Users/wangc/projects/wangc/raven/framework/Simulation.py", line 798, in run
    stepInstance.takeAstep(stepInputDict)
  File "/Users/wangc/projects/wangc/raven/framework/Steps.py", line 312, in takeAstep
    self._localTakeAstepRun(inDictionary)
  File "/Users/wangc/projects/wangc/raven/framework/Steps.py", line 466, in _localTakeAstepRun
    model.collectOutput(finishedJob,output)
  File "/Users/wangc/projects/wangc/raven/framework/Models/PostProcessor.py", line 240, in collectOutput
    self.interface.collectOutput(finishedjob,output)
  File "/Users/wangc/projects/wangc/raven/framework/PostProcessors/InterfacedPostProcessor.py", line 198, in collectOutput
    output.load(evaluation['data'], style='dict', dims=evaluation['dims'])
  File "/Users/wangc/projects/wangc/raven/framework/DataObjects/DataSet.py", line 435, in load
    self._fromDict(fileName,**kwargs)
  File "/Users/wangc/projects/wangc/raven/framework/DataObjects/DataSet.py", line 1257, in _fromDict
    self.asDataset()
  File "/Users/wangc/projects/wangc/raven/framework/DataObjects/DataSet.py", line 260, in asDataset
    return self._convertToXrDataset()
  File "/Users/wangc/projects/wangc/raven/framework/DataObjects/DataSet.py", line 1074, in _convertToXrDataset
    raise e
ValueError: setting an array element with a sequence.
```
##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### LINKED ISSUE #740: Improvement of Interface PostProcessor
--------
Issue Description
--------
##### What did you expect to see happen?
The Interface PostProcessor should take reasonable time to processing GigaByte data.

##### What did you see instead?
As observed by @alfoa, it took several hours and did not finish the preparation of the input data.

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Close #740 

Close #739 

##### What are the significant changes in functionality due to this change request?
Defect fix, and some improvement for the data preparation of InterfacePP.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/DataObjects/DataSet.py [hunk 1]
```diff
@@ -986,25 +986,30 @@ def _convertToDict(self):
...
+      # need to remove NaNs, so loop over slices
       for s,rlz in enumerate(self.sliceByIndex(self.sampleTag)):
-        # get data specific to this var for this realization (slice)
-        data = rlz[var]
-        # need to drop indexes for which no values are present
-        for index in dims:
-          data = data.dropna(index)
-          dataDict['data'][index][s] = data[index].values
-        dataDict['data'][var][s] = data.values
+        ## loop over realizations to get distinct values without NaNs
```
Comment (alfoa) [inline framework/DataObjects/DataSet.py]: performance of this?

File: tests/framework/PostProcessors/InterfacedPostProcessor/gold/HistorySetSnapShot/PrintPPpointSet_dump3.xml [hunk 2]
```diff
@@ -1,6 +0,0 @@
-<data name="PrintPPpointSet_dump3" type="Pointset"><input>y0,x0,z0</input><output>y,x,z,time</output><inputFilename>PrintPPpointSet_dump3.csv</inputFilename><metadata><SamplerType>array([u'MC', u'MC'],
```
Comment (alfoa) [inline tests/framework/PostProcessors/InterfacedPostProcessor/gold/HistorySetSnapShot/PrintPPpointSet_dump3.xml]: I can not believe we still had these files

File: tests/framework/PostProcessors/InterfacedPostProcessor/gold/historySampling/PrintPPHistorySet_dump1_0.csv [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
+0.0,3.24346738646,4.34359778869,4.15626305329
+0.0544444444444,4.54560144726,8.70873348729,4.74821381305
+0.108888888889,7.50981721821,14.9350741187,7.3944487416
+0.163333333333,12.3226460228,22.9312220448,15.3278350715
+0.217777777778,18.0422627229,26.2689310028,32.5071756938
+0.272222222222,19.4888099493,11.6874705955,49.5699281169
+0.326666666667,11.6636017187,-8.36348704689,45.4112588876
+0.381111111111,1.49496346427,-13.3433117535,35.2325304154
+0.435555555556,-4.84420933392,-12.3516934332,31.3951750587
+0.49,-7.79953251639,-10.6857759075,30.8501670182
```
Comment (alfoa) [inline tests/framework/PostProcessors/InterfacedPostProcessor/gold/historySampling/PrintPPHistorySet_dump1_0.csv]: ok

File: tests/framework/PostProcessors/InterfacedPostProcessor/gold/historySampling/PrintPPHistorySet_dump3_0.csv [hunk 4]
```diff
@@ -1,11 +1,11 @@
...
+0.0,3.24346738646,4.34359778869,4.15626305329
+0.14,10.0347739909,19.4656461422,10.972333547
+0.2,16.2793646369,26.5520948574,25.9142492987
+0.23,19.0239622191,24.85275777,37.2350331643
+0.25,19.930893694,20.0758646528,44.357188692
+0.27,19.6123489225,12.6241681416,49.2315338988
+0.3,16.4715108627,-0.0653474841513,49.9938169493
+0.33,10.9995317893,-9.08194531841,44.6878390564
+0.37,3.31875673643,-13.1772160841,36.7927086743
+0.49,-7.79953251639,-10.6857759075,30.8501670182
```
Comment (alfoa) [inline tests/framework/PostProcessors/InterfacedPostProcessor/gold/historySampling/PrintPPHistorySet_dump3_0.csv]: just order change

**Comment by alfoa** [comment]
@wangcj05

**Comment by wangcj05** [comment]
I will perform a test comparison.

**Comment by wangcj05** [comment]
@alfoa Performance of this modifications:
Test 1: 10 realizations with 160 time steps and 169 variables, using  InterfacedPostProcessor:HistorySetSampling
Devel Branch: ~27 seconds to finish
Branch in this PR: ~5 seconds to finish.

Test 2: 50 realizations with 160 time steps and 169 variables, using  InterfacedPostProcessor:HistorySetSampling
Devel Branch: ~125 seconds to finish
Branch in this PR: ~19 seconds to finish.

**Comment by alfoa** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). ***n/a***
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. ***ok***
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. ***ok***
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. ***test added***
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. ***n/a***
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. ***#740 #739***
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? ***n/a***

**Comment by alfoa** [comment]
Check list passed...

Checklist passed...

The machines TesOpenSUSE Leap 42, Test OpenSUSE Tumbleweed and Test CentOS 6 are still not functional so are considered "allowed to fail"

PR can be accepted and merged...

**Pull request merged**